JavaScript/JavaScriptSDK/SendBuffer.ts

```diff
@@ -0,0 +1,205 @@
+
+        /**
+         * Removes items from the SENT_BUFFER. Should be called on successful response from the backend. 
+         */
+        clearSent: (payload: string[]) => void;
+    }
+
+    /*
+     * An array based send buffer. 
+     */
+    export class ArraySendBuffer implements ISendBuffer {
+        private _config: ISenderConfig;
+        private _buffer: string[];
+
+        constructor(config: ISenderConfig) {
+            this._config = config;
+
+            this._buffer = [];
+        }
+
+        public enqueue(payload: string) {
+            this._buffer.push(payload);
+        }
+
+        public count(): number {
+            return this._buffer.length;
+        }
+
+        public clear() {
+            this._buffer.length = 0;
+        }
+
+        public getItems(): string[] {
+            return this._buffer;
+        }
+
+        public batchPayloads(): string {
+            if (this.count() > 0) {
+                var batch = this._config.emitLineDelimitedJson() ?
+                    this._buffer.join("\n") :
+                    "[" + this._buffer.join(",") + "]";
+
+                return batch;
+            }
+
+            return null;
+        }
+
+        public markAsSent(payload: string[]) {
+            this.clear();
+        }
+
+        public clearSent(payload: string[]) {
+            // already cleared
+        }
+    }
+
+    /*
+     * Session storege buffer holds a copy of all unsent items in the browser session storage.
+     */
+    export class SessionStorageSendBuffer implements ISendBuffer {
+        static BUFFER_KEY = "AI_buffer";
+        static SENT_BUFFER_KEY = "AI_sentBuffer";
+
+        // An in-memory copy of the buffer. A copy is saved to the session storage on enqueue() and clear(). 
+        // The buffer is restored in a constructor and contains unsent events from a previous page.
+        private _buffer: string[];
+        private _config: ISenderConfig;
+
+        constructor(config: ISenderConfig) {
+            this._config = config;
+
+            var bufferItems = this.getBuffer(SessionStorageSendBuffer.BUFFER_KEY);
+            var notDeliveredItems = this.getBuffer(SessionStorageSendBuffer.SENT_BUFFER_KEY);
+
+            this._buffer = bufferItems.concat(notDeliveredItems);
+
+            this.setBuffer(SessionStorageSendBuffer.BUFFER_KEY, this._buffer);
+            this.setBuffer(SessionStorageSendBuffer.SENT_BUFFER_KEY, []);
+        }
+
+        public enqueue(payload: string) {
+            this._buffer.push(payload);
+            this.setBuffer(SessionStorageSendBuffer.BUFFER_KEY, this._buffer);
+        }
+
+        public count(): number {
+            return this._buffer.length;
+        }
+
+        public clear() {
+            this._buffer.length = 0;
+            this.setBuffer(SessionStorageSendBuffer.BUFFER_KEY, []);
+        }
+
+        public getItems(): string[] {
+            return this._buffer;
+        }
+
+        public batchPayloads(): string {
```

Comment:
This method appears to be the same between the two buffer classes, should they share a common abstract base?